exporter/clickhouseexporter/config.go

```diff
@@ -47,6 +47,10 @@ type Config struct {
 	TableEngine TableEngine `mapstructure:"table_engine"`
 	// ClusterName if set will append `ON CLUSTER` with the provided name when creating tables.
 	ClusterName string `mapstructure:"cluster_name"`
+	// AsyncInsert if true will enable async inserts. Default is `true`.
+	// Ignored if async inserts are configured in the `endpoint` or `connection_params`.
+	// Async inserts may still be overridden server-side.
+	AsyncInsert *bool `mapstructure:"async_insert"`
```

Comment:
The default only gets applied if it's not in `endpoint`/`connection_params` AND if `cfg.AsyncInsert` is `nil`.

If it were a regular `bool`, Go would initialize the struct with `false`, and so in the config parsing, I wouldn't be able to determine if it was set to `false` by the user or by the struct's initialization.

So the code would look like:
```go
// This effectively ignores my config if I set cfg.AsyncInsert to false.
if cfg.AsyncInsert == false {
    queryParams.Set("async_insert", "true")
}
```

So if I wanted it to remain `false` in my config, it would never work, I would have to set it in `endpoint` or `connection_params`.

This is similar to the `database` config option, except in that case Go will initialize it to `""`, so the code checks for that instead of `nil`.


TL;DR: false is `false`, true is `true`, `nil` is `true`

Reply:
Okay, I think I understand your concern. However, the way configurations are parsed by the collector works in this case. It first calls [createDefaultConfig](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e870c7f0c05ec03e860df1a6a03e84dc0e308bb4/exporter/clickhouseexporter/factory.go#L32), then parses whatever configuration the user defined, overwriting any default values with given values (and leaving default values as-is if not defined by user). This is done [here](https://github.com/open-telemetry/opentelemetry-collector/blob/e8cabb785bb7bffd6c2e49f09563f132719495ea/otelcol/internal/configunmarshaler/configs.go#L41-L48), for reference.

So even if you have a `bool` with default `true`, if you set it to `true` in `createDefaultConfig`, it will only be overwritten if the user specifies `async_insert: false` in their config.

So would it work to use `bool` for your use case, or is there more reason than that?

Reply:
Hey @SpencerTorres, I was wondering if you got a chance to test this, or look into it more?